src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java

```diff
@@ -1,25 +1,110 @@
 package com.es.phoneshop.model.product;
 
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.Currency;
 import java.util.List;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.stream.Collectors;
 
 public class ArrayListProductDao implements ProductDao {
+    private long currentId;
+    private List<Product> products;
+    private final Lock productsLock = new ReentrantLock();
```

Comment:
nice attempt for the synchronization, but lets make it even better
your ReentrantLock locks everything: if one thread is reading, and you have another thread which wants to read too, the second thread will be waiting until the first is done
you can use ReadWriteLock interface instead of Lock to have 2 locks: for read and write